File: knightTravel.hs
module Knight where

-- Data Types used in program
-- Declaración de los tipos usados
-- para el contenido del Nodo
type File   = [Bool]
type Board  = [File]
type Square = (Int,Int)
type Path   = [Square]

-- Nodo del árbol de búsqueda de soluciones:
  -- Tamaño del tablero, Longitud camino construido,
  -- Estado tablero, Casilla actual caballo, Camino construido
type Node = (Int,Int,Board,Square,Path)

------------------------
-- Main Functions
------------------------
-- Comprueba si un nodo contiene, o no, un camino completo 
-- (el que recorre todo el tablero)
-- node     Nodo a comprobar su longitud
-- Devuelve True si ha recorrido todo el tablero 
fullPath :: Node -> Bool
fullPath (size,len,_,_,_) = len == (size*size)

-- Recibe un nodo y devuelve lista de compleciones de dicho nodo, 
-- nodos en los que se visita una nueva casilla válida 
-- node     Nodo a revisar sus saltos
-- Devuelve la lista de nodos [Node] no visitados y
  -- dentro del tablero
validJumps :: Node -> [Node]
validJumps (size,len,board,(c,r),path) =
  return (jumps (size,len,board,(c,r),path)) -- listado de saltos
    where       
      -- añade las casillas válidas instanciando cada nodo
      return []     = [] -- si no existen saltos
      return (x:xs) = 
        [(size,len+1,markSquare board x,x,addSquare path x)] ++ return xs

-- Función Principal 
-- size        tamaño del tablero 
-- initSquare  casilla desde la que el caballo inicia recorrido y
-- Devuelve el camino que recorre todo el tablero comenzando
   -- por la casilla indicada. 
   -- Si no solución, camino devuelto igual a vacío.
knightTravel :: Int -> Square -> Path
knightTravel size initSquare 
    -- nodo inicial no se encuentra dentro del tablero
  | not (inBoard size initSquare) = [] 
  | otherwise                     = getFirstPath nodeList
  where
    --nodo inicial
    node = (size, 1, markSquare (initBoard size) initSquare,
            initSquare, addSquare [] initSquare)
    --lista de nodos con soluciones
    nodeList = bt (fullPath) (validJumps) node

-- Backtracking scheme
-- isSol  función que comprueba si un nodo es, o no, solución. [fullPath]
-- comp   función que, dado un nodo, devuelve sus compleciones. [validJumps]
-- node   nodo a partir del cual explorar
-- Devuelve lista de TODOS los nodos que contengan soluciones al problema
bt :: (a -> Bool) -> (a -> [a]) -> a -> [a]
bt isSol comp node
  | isSol node = [node]
  | otherwise  = concat (map (bt isSol comp) (comp node))

-------------------------
-- Auxiliary Functions
-------------------------
-- Estado inicial del tablero 
-- size    tamaño del tablero 
-- Devuelve tablero con todas casillas a True
initBoard :: Int -> Board
initBoard size = replicate size (replicate size True)

-- Marca una casilla como visitada
-- board tablero a marcar
-- (c,r) casilla a marcar
-- Devuelve el tablero con la casilla marcada
markSquare :: Board -> Square -> Board
markSquare board (c,r) = 
    take (r-1) board ++ 
    [( take (c-1) (board !! (r-1)) ++ [False] ++ drop (c) (board !! (r-1)) )]
    ++ drop (r) board 

-- Revisa si la casilla está dentro del tablero
-- size  tamaño del tablero
-- (c,r) casilla a revisar
-- Devuelve True si se encuentra en el tablero
inBoard :: Int -> Square -> Bool
inBoard size (c,r)
  |  (c > size) || (r > size) = False
  |  (c < 1)    || (r < 1)    = False
  |  otherwise                = True

-- Añade una casilla, al final del camino creado
-- hasta ahora
-- path  camino almacenado
-- sq    casilla a añadir
-- Devuelve el camino actualizado
addSquare :: Path -> Square -> Path
addSquare path sq = path ++ [sq]

-- Comprueba si el salto, para la casilla, puede realizarse,
-- revisando que se encuentra dentro del tablero,
-- y no se encuentra visitada (false)
-- size  tamaño del tablero
-- board tablero a revisar
-- (c,r) casilla a revisar
-- Devuelve el camino actualizado
validJump :: Int ->  Board -> Square -> Path
validJump size board (c,r) 
  | inBoard size (c,r) && (board !! (r-1)) !! (c-1) = [(c,r)]
  | otherwise                                       = []

-- Revisa un nodo y todos sus saltos válidos,
-- haciendo el cálculo con cada uno de los posibles 
-- saltos del caballo
-- node  nodo a revisar
-- Devuelve un camino de todas las casillas válidas
jumps :: Node -> Path
jumps (size,len,board,(c,r),path) = 
  concat (map (validJump size board) [(x1,y1),
                                     (x2,y2),
                                     (x3,y2),
                                     (x4,y1),
                                     (x4,y3),
                                     (x3,y4),
                                     (x2,y4),
                                     (x1,y3)])
  -- saltos posibles del caballo
    where
      x1 = c + 2
      y1 = r + 1
      x2 = c + 1
      y2 = r + 2
      x3 = c - 1
      y3 = r - 1
      x4 = c - 2
      y4 = r - 2

-- Consigue el camino del primer nodo
-- [node]  nodo/s del que obtener el camino
-- Devuelve el camino del primer nodo si hay,
--        si no devuelve el camino vacio
getFirstPath :: [Node] -> Path
getFirstPath nodeList
  | nodeList /= [] = getPath (nodeList!!0)
  | otherwise      = []
    where
      getPath (_,_,_,_,path) = path

File: knightTravel.pl
%-------------------------------
% Representación de términos
%-------------------------------
/**
 * - Casilla: F y R enteros que representan el 
 * número de columna (file) y de fila (rank) respectivamente
 * s(F,R)
 * - Nodo: nodo del espacio de búsqueda de soluciones (node) se 
 * representa por la función con cinco argumentos, que a su vez 
 * son términos Prolog:
 * node(Tamanyo_Board,Long_Path_recorrido,Board,Square_actual,Path_recorrido).
 */

/** 
 * Construye un tablero vacío de nxn (N) en una ejecución
 * con todas las casillas a True.
 */
tableroVacio(N,Board):- ct_aux(1,N,Board).

ct_aux(M,N,[]):- M > N, !.
ct_aux(M,N,[File|RF]):-
    columnaVacia(N,File), M1 is M+1, ct_aux(M1,N,RF).

columnaVacia(1,[true]):- !.
columnaVacia(N,[true|CV]):- N1 is N-1, columnaVacia(N1,CV).

/** 
 * Comprueba si en un nodo del espacio de búsqueda hay, o no,
 * una solución, si contiene, o no, un camino completo, 
 * a partir del tamaño N y la longitud del camino L. 
 */
fullPath(node(N,L,_,_,_)):- (N*N) =:= L.

/** 
 * Situa el nodo inicial (comprobando que esté en el tablero)
 * a partir del tamaño del tablero, N, y una casilla desde 
 * la que el caballo inicia su recorrido, s(F,R), en un tablero vacío.
 */
initBoard(N,s(F,R),node(N,1,Board,s(F,R),[s(F,R)])):-
    tableroVacio(N,Board0),
    inBoard(N,s(F,R)),!,
    visitSquare(s(F,R),Board0,Board).

/** 
 * Comprueba si la casilla s(F,R) está dentro de un tablero 
 * de tamaño N.
 */
inBoard(N,s(F,R)):- 
    N >= F, N >= R,
    F >= 1, R >= 1.

/** 
 * Dada una casilla (s(F,R)), la visita (cambia su
 * valor a falso), en un tablero dado (Board),
 * generando el nuevo.
 */
visitSquare(s(F,R),Board,BoardSol):- markRank(Board,R,BoardSol,F).

/**
 * Función principal, entero representa el tamaño del tablero, 
 *                    casilla inicial del caballo 
 * Devuelve camino que recorre todo el tablero comenzando
 *                     por la casilla indicada. 
 *                     Si no hubiera solución, camino devuelto == vacío.
 */
knightTravel(N,s(F,R)):-
    initBoard(N,s(F,R),node(N,1,Board,s(F,R),[s(F,R)])),
    recorrerBT(node(N,1,Board,s(F,R),[s(F,R)])),!. 
% si se encuentra camino no se ejecuta el camino vacío
knightTravel(_,_):-
    write([]).

/**
 * Búsqueda de soluciones en profundidad con vuelta
 * atrás con una llamada recursiva, saltando de nodo
 * en nodo. 
 * Condición de parada (con Corte), encontrar un nodo 
 * solución (fullPath).
 */
recorrerBT(node(N,L,_,_,Path)):-
    fullPath(node(N,L,_,_,_)),
    write(Path),!.
recorrerBT(node(N,L,Board,s(F,R),Path)):-
    jump(node(N,L,Board,s(F,R),Path),NNuevo),
    recorrerBT(NNuevo).

/** 
 * Comprueba que haya nodos sucesores a partir de Node y
 * construye los movimientos posibles(knightMove), 
 * de uno en uno, comprobando si son casillas 
 * válidas (validSquare) y por lo tanto movimientos
 * válidos que generan un nodo sucesor (NNode).
 */
jump(node(N,L,Board,s(F,R),Path), node(N,NL,BoardNuevo,s(NF,NR),NPath)):-
    knightMove(F,R,NF,NR),
    validSquare(N,s(NF,NR),Board,BoardNuevo),
    NL is L+1, 
    append(Path,[s(NF,NR)],NPath).

/** 
 * Comprueba en el tablero (Board) de tamaño N, 
 * que puede ir a casilla s(F,R), 
 *    porque está en el tablero (inBoard) y
 *    está libre (freeSquare); 
 *    Coloca (pone a false la casilla en 
 * el tablero con visitSquare)
 * el caballo en posición s(F,R), construyendo el nuevo 
 * tablero (BoardNuevo).
 */
validSquare(N,s(F,R),Board,BoardNuevo):-
    inBoard(N,s(F,R)),
    freeSquare(s(F,R),Board),
    visitSquare(s(F,R),Board,BoardNuevo). 

%%%%%%%%%%%%%%%%%%%%%%%
% Auxiliar predicates %
%%%%%%%%%%%%%%%%%%%%%%%
/** 
 * Marca la casilla como false, generandola  
 * en un nuevo tablero, tanto con las mismas filas
 * (markRank), como columnas (markFile)
 */
markRank([],_,[],_):- !.
markRank([H|T],1,[HF|Rest],F):-
    markFile(H,F,HF),
    markedRank(T,Rest).
markRank([H|T],R,[H|Rest],F):-
    R > 1,
    NR is R-1,
    markedFile(H,H),
    markRank(T,NR,Rest,F).
      
markedRank([],_):-!.
markedRank([H|T],[H|Rest]):-
    markedFile(H,H),
    markedRank(T,Rest).

% una vez que se ha marcado el File se continua igual
% con el resto del Board  
markedFile([],_):-!.
markedFile([H|T],[H|Rest]):-
    markedFile(T,Rest).

markFile([],_,_):-!.
markFile([true|T],1,[false|Rest]):-
    markedFile(T,Rest).
markFile([H|T],F,[H|Rest]):-
    F > 1,
    NF is F-1,
    markFile(T,NF,Rest).
/** 
 * Comprueba que la casilla está libre (true) 
 * iterando por las filas (rankFree) y 
 * columnas (fileFree) del tablero
 */
freeSquare(s(F,R),Board):- rankFree(Board,R,F).

rankFree([],_,_):- !.
rankFree([H|_],1,F):- fileFree(H,true,F),!. 
rankFree([_|Rest],R,F):-
    NR is R-1, rankFree(Rest,NR,F).

fileFree([],_,_):- !.
fileFree([H|_],B,1):- H==B,!.
fileFree([_|Rest],B,F):-
    NF is F-1, fileFree(Rest,B,NF).

% Posibles movimientos del caballo
knightMove(F,R,NF,NR):- NF is F+2, NR is R+1.
knightMove(F,R,NF,NR):- NF is F+2, NR is R-1.
knightMove(F,R,NF,NR):- NF is F+1, NR is R+2.
knightMove(F,R,NF,NR):- NF is F+1, NR is R-2.
knightMove(F,R,NF,NR):- NF is F-1, NR is R+2.
knightMove(F,R,NF,NR):- NF is F-1, NR is R-2.
knightMove(F,R,NF,NR):- NF is F-2, NR is R+1.
knightMove(F,R,NF,NR):- NF is F-2, NR is R-1.